### PR TITLE
Python API for inference model saving/load

### LINKED ISSUE #5019: Need Python API to save models in refactored Paddle
Currently, `variables` can be saved by `save op` (in review: https://github.com/PaddlePaddle/Paddle/pull/4602) and `ProgramDesc` can also be easily serialised and saved. However, we don't have a unified Python API to invoke them.

### PR DESCRIPTION
**Pull request by Canpio**
fixes #5019 
depends on https://github.com/PaddlePaddle/Paddle/pull/4602

### CODE REVIEW

File: python/paddle/v2/framework/dump_load.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        os.makedirs(folder_path)
+
+    block = program.global_block()
+    # add feed/fetch vars and ops to block
+    feed_order = utility.add_feed_components(block, feeded_var_names, "feed")
+    fetch_order = utility.add_fetch_components(block, target_vars, "fetch")
+    # label target op
+    for var in target_vars:
+        var.op.mark_as_target()
+
```
Comment (reyoung) [inline python/paddle/v2/framework/dump_load.py]: logic is too complex below.

```python
with open(model_file_name, 'w') as f:
  pickle.dump({
    'program_desc': program.desc.serialize_to_string(),
    'feed_order': feed_order,
    'fetch_order': fetch_order
  }, f, -1)
```

File: python/paddle/v2/framework/dump_load.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            buf.write(binary_str)
+            tarinfo = tarfile.TarInfo(name=to_dump[1])
+            buf.seek(0)
+            tarinfo.size = len(buf.getvalue())
+            tar.addfile(tarinfo=tarinfo, fileobj=buf)
+
+    # Build another program to hold save_ops
+    _dump_all_persistable_vars_(
+        folder_path=folder_path, program=program,
+        exempt_list={"feed", "fetch"})
```
Comment (reyoung) [inline python/paddle/v2/framework/dump_load.py]: the variable name will be changed in different models, `exempt_list` cannot just set to `feed`/`fetch`

File: python/paddle/v2/framework/dump_load.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    if program is None:
+        program = framework.g_program
+    dump_program = framework.Program()
+    dump_block = dump_program.global_block()
+    if exempt_list is None:
+        exempt_list = {}
+
+    save_op_inputs = []
+    for var in program.global_block().vars.itervalues():
+        if var.desc.persistable() and not var.name in exempt_list:
```
Comment (reyoung) [inline python/paddle/v2/framework/dump_load.py]:
```python
if var.persistable and var.type == core.VarDesc.VarType.LOD_TENSOR:
```

Comment (Canpio) [reply python/paddle/v2/framework/dump_load.py]: done

File: python/paddle/v2/framework/dump_load.py (hunk 1 above)
Comment (Canpio) [reply python/paddle/v2/framework/dump_load.py]: Impressive!

File: python/paddle/v2/framework/dump_load.py (hunk 2 above)
Comment (Canpio) [reply python/paddle/v2/framework/dump_load.py]: By 
```python
if var.persistable and var.type == core.VarDesc.VarType.LOD_TENSOR:
```
there is no necessary to use the `exempt_list`

File: python/paddle/v2/framework/framework.py [hunk 4]
```diff
@@ -251,6 +251,8 @@ def find_name(var_list, name):
                 self.desc.set_output(out_proto.name, out_argu_names)
 
         if attrs is not None:
+            if not isinstance(attrs, dict):
+                raise ValueError("'attrs' should be a dict.")
```
Comment (reyoung) [inline python/paddle/v2/framework/framework.py]: raise TypeError

File: python/paddle/v2/framework/io.py [hunk 5]
```diff
@@ -1,11 +1,12 @@
 import os
+import pickle
```
Comment (reyoung) [inline python/paddle/v2/framework/io.py]: import cPickle as pickle

Comment (reyoung) [reply python/paddle/v2/framework/io.py]: pickle may 10x slower than cPickle

Comment (Canpio) [reply python/paddle/v2/framework/io.py]: Thanks! Done.

File: python/paddle/v2/framework/framework.py (hunk 4 above)
Comment (Canpio) [reply python/paddle/v2/framework/framework.py]: Done.

**Pull request merged**